crates/sui-bcs-json/src/lib.rs

```diff
@@ -0,0 +1,648 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use ::safer_ffi::prelude::*;
+use anyhow::{anyhow, bail, Error};
+use bcs::{from_bytes as bcs_from_bytes, to_bytes as bcs_to_bytes};
+use move_core_types::u256::U256;
+use safer_ffi::libc::size_t;
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value, Value::Array};
+use sui_types::{
+    base_types::{ObjectDigest, ObjectID, SequenceNumber, SuiAddress},
+    transaction::{Argument, CallArg, ProgrammableTransaction, TransactionData, TransactionKind},
+};
+
+/// Deserializes the data to given type T, and then serializes it as a JSON value
+fn to_json<T: Serialize + for<'a> Deserialize<'a>>(data: &[u8]) -> Result<Value, Error> {
+    let bcs: T = bcs_from_bytes(data)?;
+    Ok(json!(bcs))
+}
+
+/// Deserializes the string data to a given type T, and then serializes it as a BCS array
+fn to_bcs<T: Serialize + for<'a> Deserialize<'a>>(data: &str) -> Result<Vec<u8>, Error> {
+    let bcs: T = serde_json::from_str(data)?;
+    Ok(bcs_to_bytes(&bcs)?)
+}
+
+/// Deserializes a BCS array to an integer,
+/// and then formats the integer as a string,
+/// and serializes it to a json if integer_as_string is set to true.
+///
+/// By default, u8, u16, u32 are serialized as JSON numbers,
+/// while u64, u128, and U256 are serialized as JSON strings.
+fn string_integer_to_json(typename: &str, data: &[u8]) -> Result<Value, Error> {
+    match typename {
+        "u8" => {
+            let bcs: u8 = bcs_from_bytes(data)?;
+            Ok(json!(bcs))
+        }
+        "u16" => {
+            let bcs: u16 = bcs_from_bytes(data)?;
+            Ok(json!(bcs))
+        }
+        "u32" => {
+            let bcs: u32 = bcs_from_bytes(data)?;
+            Ok(json!(bcs))
```

Comment:
why are these not converted to strings?

Reply:
In the TS reimplementation of the BCS library, it was decided to only keep `u64,u128,u256` as strings, thus I thought we should be consistent across our implementations.